std/file.d

```diff
@@ -670,20 +651,200 @@ version(Windows) private void writeImpl(const(char)[] name, const(FSChar)* namez
         cenforce(h != INVALID_HANDLE_VALUE, name, namez);
     }
+    writeToOpenFile(h, buffer, name, namez);
+}
+
+private void writeToOpenFile(FD)(FD fd, const void[] buffer, const(char)[] name,
+                                 const(FSChar)* namez) @trusted
+{
     immutable size = buffer.length;
     size_t sum, cnt = void;
-    DWORD numwritten = void;
+
     while (sum != size)
     {
-        cnt = (size - sum < 2^^30) ? (size - sum) : 2^^30;
-        WriteFile(h, buffer.ptr + sum, cast(uint) cnt, &numwritten, null);
-        if (numwritten != cnt)
+        cnt = size - sum < 2^^30 ? size - sum : 2^^30;
+
+        version(Posix)
+            immutable numWritten = core.sys.posix.unistd.write(fd, buffer.ptr + sum, cnt);
+        else version(Windows)
+        {
+            DWORD numWritten = void;
+            WriteFile(fd, buffer.ptr + sum, cast(uint) cnt, &numWritten, null);
+        }
+        else
+            static assert(0, "Unsupported OS");
+
+        if (numWritten != cnt)
             break;
-        sum += numwritten;
+        sum += numWritten;
     }
-    cenforce(sum == size && CloseHandle(h), name, namez);
+
+    version(Posix)
+        cenforce(sum == size && core.sys.posix.unistd.close(fd) == 0, name, namez);
+    else version(Windows)
+        cenforce(sum == size && CloseHandle(fd), name, namez);
+    else
+        static assert(0, "Unsupported OS");
 }
 
+
+/++
+    Creates a file with a random name in $(LREF tempDir) and returns the name
+    of the file. If data is passed in, then the file is written with that data;
+    otherwise, it's empty.
+
+    The idea is that this creates a temporary file for testing or whatever
+    other purpose a temporary file might be needed for. However, it will only
+    be deleted if it's explicitly deleted or if the OS decides to delete it as
+    some OSes do with files in temp directories on startup or shutdown. So,
+    unlike $(REF tmpfile,std,stdio), the file has a name, and it can have all
+    of the things done to it that one might typically do with a file (including
+    reopening it after closing it).
+
+    The file is created with R/W permissions. On POSIX systems, the permissions
+    are restricted to the current user, though the effective permissions are
+    modified by the process' umask in the usual way.
+
+    Params:
+        prefix = Prefix to the random portion of the file name.
+        suffix = Suffix to the random portion of the file name (e.g. for a file
+                 extension).
+        buffer = Data to populate the file with.
+
+    Returns:
+        The name of the generated file.
+
+    Throws:
+        $(LREF FileException) if it fails to create the file.
+
+    See_Also:
+        $(REF tmpfile,std,stdio)
+  +/
+string createTempFile(const void[] buffer = null) @safe
+{
+    return createTempFile(null, null, buffer);
+}
+
+/// Ditto
+string createTempFile(const(char)[] prefix, const(char)[] suffix, const void[] buffer = null) @trusted
```

Comment:
Why not `string` or using templates?

Reply:
If it were `string` and the prefix/suffix weren't already immutable then it would be necessary to create copies of them to pass to the function. What benefit would immutability provide in exchange?

If the prefix/suffix were template parameters then it would be impossible for them to be specified at runtime, so I think the question is whether using templates gives some benefit that justifies that limitation.